### DISCUSSION TITLE
Losing object reference after lightsleep ESP32

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#11877] [Using MicroPython]
Just sitting in a while loop waiting for an interrupt from the
RFM96W, but also waking-up from lightsleep every 55 seconds.
I find that I have to call `lora = SX127x(spi, pins = cnfg, parameters = loracnfg)`
each time around the while loop.

If I put `lora = SX127x(spi, pins = cnfg, parameters = loracnfg)`
outside of this while loop it runs once a second printing `b'no msg'`, ie no more lightsleep
```py
msg = b'no msg'

while True:
    lora = SX127x(spi, pins = cnfg, parameters = loracnfg) #  needs to be done each loop
    lora.receive()

    print('going to lightsleep for 55 seconds')

 #  hold the GPIO state during lightsleep
    reset_pin.init(hold=True)

    lightsleep(55000) # go to lightsleep for 55 seconds

 #  unhold the GPIO state after lightsleep
    reset_pin.init(hold=False)

    Rx = lora.read_payload()

    if Rx:
        try:
            msg,rssi,snr = Rx,lora.packet_rssi(),lora.packet_snr()
            print('msg', msg, ' rssi', rssi, ' snr', int(snr))
        except Exception as error:
            pass
    utime.sleep(1)
        
 #  process the msg
```
I have checked that reset_pin does stay high during the entire while loop.

When lora is defined outside of this loop I print out its object location.  Then print out it's location before and after the lightsleep it is apparent that it's location changes after it comes out of lightsleep when triggered by wake_on_ext1. The wake_on_ext1 occurs when a RFM96W rxDone interrupt happens.

From the docs:
> A lightsleep has full RAM and state retention. Upon wake execution is resumed from the point where the sleep was requested, with all subsystems operational.

Why does the object location for lora change?

**Comment by davefes** [comment]
I can not reproduce this problem so I must be doing something else wrong.

**Reply by glenn20** [reply]
Normally, lightsleep will powerdown various peripherals. I don't know about your lora device, but it may well be that your peripheral is losing power on the ligthsleep (or potentially, may not be determinsticallly powered - leading to inconsistent behaviour after the ligthsleep. I have observed similar inconsistency in the behaviour of the wifi peripheral after a ligthsleep - sometimes need to be re-initialised, and sometimes not).

See https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/sleep_modes.html:
>  In Light-sleep mode, the digital peripherals, most of the RAM, and CPUs are clock-gated and their supply voltage is reduced. Upon exit from Light-sleep, the digital peripherals, RAM, and CPUs resume operation and their internal states are preserved.

**Reply by davefes** [reply]
Thanks for the reply.  The RFM96W has it's own supply and I made sure that it was not being reset.  I suspect that it might have been a programming error.
